<launch>
  <arg name="use_recovery_behaviors" default="true"/>
  
  <!--<include file="$(find turtlebot_navigation)/launch/includes/velocity_smoother.launch.xml"/>-->

  <arg name="vel_topic" default="/robot0/cmd_vel"/>
  <arg name="holonomic" default="true"/>
  <arg name="projection_operator" default="false"/>

  <node name="egocircle_node" pkg="egocircle" type="ego_circle">
      <param name="base_frame_id" value="robot0"/>
      <param name="odom_frame_id" value="map_static"/>
      <remap from="/scan" to="/robot0/laser_0"/>
  </node>
  
  <arg name="global_planning_freq" default="$(optenv GM_PARAM_GLOBAL_PLANNING_FREQ 1)"/>
  <arg name="controller_freq" default="$(optenv GM_PARAM_CONTROLLER_FREQ 5)"/>
  <arg name="planning_inflated" default="$(optenv GM_PARAM_PLANNING_INFLATED false)"/>
  <arg name="far_feasible" default="$(optenv GM_PARAM_FAR_FEASIBLE true)"/>
  <arg name="full_fov" default="$(optenv GM_PARAM_FULL_FOV false)"/>
  <arg name="inf_ratio" value="1.2" unless="$(arg planning_inflated)"/>
  <arg name="inf_ratio" value="0.0" if="$(arg planning_inflated)"/>
  <arg name="reduction_threshold" default="$(optenv GM_PARAM_REDUCTION_THRESHOLD 3.1415926)"/>
  <arg name="reduction_target" default="$(optenv GM_PARAM_REDUCTION_TARGET 3.1415926)"/>

  <arg name="debug" default="false"/>

  <arg name="debug_prefix" value="" unless="$(arg debug)"/> <!-- valgrind - -tool=callgrind -->
  <arg name="debug_prefix" value="xterm -e gdb -ex run --args" if="$(arg debug)"/>

  <!-- xterm -e gdb -ex run - -args -->
  
  
  <env name="ROSCONSOLE_CONFIG_FILE" value="$(find nav_configs)/config/custom_teb_rosconsole.conf"/>
  

  <node pkg="move_base_virtual" type="move_base_virtual" respawn="false" name="move_base_virtual" output="screen" clear_params="true" launch-prefix="$(arg debug_prefix)">
      <rosparam file="$(find potential_gap)/config/common_costmap_params.yaml" command="load" ns="global_costmap"/>
      <rosparam file="$(find potential_gap)/config/common_costmap_params.yaml" command="load" ns="local_costmap"/>
      <rosparam file="$(find potential_gap)/config/local_costmap_params.yaml" command="load"/>
      <rosparam file="$(find potential_gap)/config/global_costmap_params.yaml" command="load"/>

      <rosparam file="$(find nav_configs)/config/local_planner_common_params.yaml" command="load" ns="PotentialGapPlanner"/>
      <rosparam file="$(find nav_configs)/config/move_base_params.yaml" command="load"/>
      <rosparam file="$(find nav_configs)/config/global_planner_params.yaml" command="load" ns="GlobalPlanner"/>

      <param name="recovery_behavior_enabled" value="false" unless="$(arg use_recovery_behaviors)"/>

      <param name="controller_frequency" value="$(arg controller_freq)" />
      <param name="planner_frequency" value="$(arg global_planning_freq)"/>

      <!--<remap from="cmd_vel" to="navigation_velocity_smoother/raw_cmd_vel"/>-->

      <remap from="cmd_vel" to="$(arg vel_topic)"/>
      <remap from="/odom" to="/robot0/odom"/>
      
      <param name="PotentialGapPlanner/map_frame_id" value="known_map"/>
      <param name="PotentialGapPlanner/odom_frame_id" value="map_static"/>
      <param name="PotentialGapPlanner/robot_frame_id" value="robot0"/>
      <param name="PotentialGapPlanner/sensor_frame_id" value="robot0_laser_0"/>

      <param name="PotentialGapPlanner/holonomic" value="$(arg holonomic)"/>
      <param name="PotentialGapPlanner/projection_operator" value="$(arg projection_operator)"/>
      <param name="PotentialGapPlanner/planning_inflated" value="$(arg planning_inflated)"/>
      <param name="PotentialGapPlanner/inf_ratio" value="$(arg inf_ratio)"/>
      <param name="PotentialGapPlanner/full_fov" value="$(arg full_fov)"/>
      <param name="PotentialGapPlanner/far_feasible" value="$(arg far_feasible)"/>
      <param name="PotentialGapPlanner/reduction_threshold" value="$(arg reduction_threshold)"/>
      <param name="PotentialGapPlanner/reduction_target" value="$(arg reduction_target)"/>


      <!-- <param name="PotentialGapPlanner/v_lin_x_const" value="0.0"/>
      <param name="PotentialGapPlanner/v_lin_y_const" value="0.0"/>
      <param name="PotentialGapPlanner/k_drive_x" value="5.5"/>
      <param name="PotentialGapPlanner/k_drive_y" value="5.5"/>
      <param name="PotentialGapPlanner/planning_inflated" value="true"/> -->

      <!-- Approach-specific parameters/remappings/etc -->
      <param name="base_local_planner" value="potential_gap/PotentialGapPlanner" />
      <rosparam file="$(find potential_gap)/config/potentialgap_local_planner_params_holonomic.yaml" command="load"/>

      <!-- Disable the local costmap observation sources if they will not be used -->
      <!--<param name="local_costmap/obstacle_layer/observation_sources" value="false" if="$(arg egocircle)"/>-->
  </node>





</launch>
